Bluetooth: Tester: Use `BT_L2CAP_SEG_RECV` (experimental variant)

alwa-nordic opened the pull request:
This PR is an alternative to https://github.com/zephyrproject-rtos/zephyr/pull/80302 and https://github.com/zephyrproject-rtos/zephyr/pull/80883.

This version introduces an experimental function `bt_l2cap_ecred_chan_reconfigure_explicit`.

We need the fix for the original issue downstream in NCS without delay to proceed with a release. We plan to apply this version downstream. This PR is to upstream this experimental version. But we can alternatively synchronize the forks on a different solution later.

sjanc committed:
Bluetooth: Host: L2CAP: Fix seg_recv call on SDU overflow

This fix calling seg_recv() callback being called even though channel
is being disconnected due to SDU overflow.

Signed-off-by: Szymon Janc <szymon.janc@codecoup.pl>

jhedberg commented on include/zephyr/bluetooth/l2cap.h:
```diff
@@ -557,6 +558,27 @@ int bt_l2cap_ecred_chan_connect(struct bt_conn *conn,
+ *  @param chans Array of channel objects. Null-terminated. Elements after the
+ *               first 5 are silently ignored.
+ *  @param mtu Channel MTU to reconfigure to.
+ *  @param mtu Channel MPS to reconfigure to.
```
s/mtu/mps/

I'd guess this is why the Documentation check is failing.

Thalley commented on include/zephyr/bluetooth/l2cap.h:
```diff
@@ -557,6 +558,27 @@ int bt_l2cap_ecred_chan_connect(struct bt_conn *conn,
 int bt_l2cap_ecred_chan_reconfigure(struct bt_l2cap_chan **chans, uint16_t mtu);
 
+/** @brief Reconfigure Enhanced Credit Based L2CAP channels
+ *
+ *  Experimental API to reconfigure with explicit MPS and MTU values.
+ *
```
@kartben do we have any good way of indicating that a specific function should be considered experiemental? Or just like this?
Is `@version` something we use for functions, or just API groups?

Thalley commented on include/zephyr/bluetooth/l2cap.h:
```diff
@@ -557,6 +558,27 @@ int bt_l2cap_ecred_chan_connect(struct bt_conn *conn,
+ *  Once reconfiguration is completed each channel reconfigured() callback will
+ *  be called. MTU cannot be decreased on any of provided channels.
+ *
+ *  Available only when @kconfig{CONFIG_BT_L2CAP_RECONFIGURE_EXPLICIT} is
+ *  enabled.
```
```suggestion
 *  @kconfig_dep{CONFIG_BT_L2CAP_RECONFIGURE_EXPLICIT}
```

Thalley commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -3035,6 +3036,105 @@ int bt_l2cap_ecred_chan_reconfigure(struct bt_l2cap_chan **chans, uint16_t mtu)
+	uint8_t ident;
+	int i;
+
+	LOG_DBG("chans %p mtu 0x%04x", chans, mtu);
```
```suggestion
	LOG_DBG("chans %p mtu 0x%04x mps 0x%04x", chans, mtu, mps);
```

Thalley commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -3035,6 +3036,105 @@ int bt_l2cap_ecred_chan_reconfigure(struct bt_l2cap_chan **chans, uint16_t mtu)
+		/* MPS is not allowed to decrease when reconfiguring multiple channels.
+		 * Core Specification 3.A.4.27 v6.0
+		 */
+		if (multiple_chan && mps < BT_L2CAP_LE_CHAN(chans[i])->rx.mps) {
+			return -EINVAL;
+		}
```
If you do 
```suggestion
		if (i > 0 && mps < BT_L2CAP_LE_CHAN(chans[i])->rx.mps) {
			return -EINVAL;
		}
```

Then you can omit the `multiple_chan` variable. It is also OK as is, as that is quite explicit and easy to read

alwa-nordic replied:
I'll keep it as is then. I clarified it like this while explaining the code to someone, so I think it is good to have.

Thalley commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -3035,6 +3036,105 @@ int bt_l2cap_ecred_chan_reconfigure(struct bt_l2cap_chan **chans, uint16_t mtu)
+		return -EINVAL;
+	}
+
+	if (mps > BT_L2CAP_RX_MTU) {
+		return -EINVAL;
+	}
```
MTU range check? 

Is there a minimum MPS value?

alwa-nordic replied:
Good point. I think it's 23, but it's worded weirdly:

> L2CAP implementations shall support a minimum MPS of 23 octets and may support an MPS up to 65533 octets.

[Core 3.A.4.22 v6.0](https://www.bluetooth.com/wp-content/uploads/Files/Specification/HTML/Core-60/out/en/host/logical-link-control-and-adaptation-protocol-specification.html#:~:text=L2CAP%20implementations%20shall%20support%20a%20minimum%20MPS%20of%2023%20octets%20and%20may%20support%20an%20MPS%20up%20to%2065533%20octets.)

What does "support" mean? Did they mean to say simply "Allowed range is 23 to 65533."? I don't understand why the spec writes make everything so difficult.

Thalley replied:
It is worded weirdly, but I think it's OK to assume that any values below 23 should be rejected. 

Do we have any checks if we receive a request from a peer with MPS < 23?

sjanc replied:
23 is for LE Credits Based connections,  for Enhanced Credit Based Connections min MPS is 64 

also, being super strict on spec,  reconfiguration is procedure defined for Enhanced Credit Based only... 🫠

alwa-nordic replied:
> 23 is for LE Credits Based connections, for Enhanced Credit Based Connections min MPS is 64
> 
> also, being super strict on spec, reconfiguration is procedure defined for Enhanced Credit Based only... 🫠

64 it is then. AFAIK, we don't have any way to tell whether a `bt_l2cap_chan` is in LE Credit Based Flow Control mode or Enhanced Credit Based Flow Control mode, so I won't be adding a check in this PR. But we should make a TODO.

Thalley commented on tests/bluetooth/tester/nrf5340_hci_ipc.conf:
This file is is applied to the network core build, and not the application itself. The changes here should be reverted

alwa-nordic replied:
@sjanc I kept these changes to be on the safe side downstream, since I know PTS passes with this applied. But are they necessary or intentional?

sjanc replied:
those were to make sure same configuration is used across boards,  nrf5340_hci_ipc.conf was modified since hci_ipc defaults to same values anyway (but maybe we should left this as is to be on the safe side if defaults changes)

alwa-nordic replied:
> This file is is applied to the network core build, and not the application itself. 

This seems to conflict with the comment at the top of this file. @sjanc, what do you make of this? Can I revert this change?

sjanc replied:
ah wait, I confused configs

nrf5340_hci_ipc.conf is for app core
nrf5340_hci_ipc_cpunet.conf is for net core

so this moves app config to prj.conf and thus removing them from nrf5340_hci_ipc.conf,  I'd let it as is in this PR (to minimize build differences across boards)

Thalley replied:
Ah, I misunderstood the files then :)

Thalley commented on tests/bluetooth/tester/prj.conf:
The changes to this file increases the memory requirements for all devices. We don't need e.g. CONFIG_BT_MAX_CONN=3 for non-LE audio, and this will make it harder to fit the BT tester on resource constrained devices

sjanc replied:
this increases RAM usage from ~72k to 77k (default non-le audio build on nRF52840)

I don't think this is a big issue to simplify configs in such case, but I won't object if others prefer not to

Thalley commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -3035,6 +3036,105 @@ int bt_l2cap_ecred_chan_reconfigure(struct bt_l2cap_chan **chans, uint16_t mtu)
+		return -EINVAL;
+	}
+
+	if (mps < 64 || mps > BT_L2CAP_RX_MTU) {
+		return -EINVAL;
+	}
```
`64` is OK for now since it's only used a single place, but I really do prefer if such MIN/MAX values are defined. That way we can also document in the function that `mps` shall be between `BT_L2CAP_MPS_MIN` and `BT_L2CAP_RX_MTU`

What's also is a bit interesting is that `BT_L2CAP_RX_MTU` may be less than 64...

alwa-nordic replied:
> `64` is OK for now since it's only used a single place, but I really do prefer if such MIN/MAX values are defined. That way we can also document in the function that `mps` shall be between `BT_L2CAP_MPS_MIN` and `BT_L2CAP_RX_MTU`

You are right. I will add a define in `l2cap.h`.

> What's also is a bit interesting is that `BT_L2CAP_RX_MTU` may be less than 64...

Oh. We should update the config defaults  when ECRED is enabled and add a build assert for that as well. Let's do that in a separate PR, in case it causes problems.

alwa-nordic commented:
I removed d8346a9915f6e1d9becebc10055a7a007833f803 "Unify L2CAP and ACL bufs configuration". By the commit message, it look like it's not needed for this PR. And it was problematic in our downstream process.

sjanc commented:
> I removed [d8346a9](https://github.com/zephyrproject-rtos/zephyr/commit/d8346a9915f6e1d9becebc10055a7a007833f803) "Unify L2CAP and ACL bufs configuration". By the commit message, it looks like it's not needed for this PR. And it was problematic in our downstream process.

it is needed for L2CAP tests to pass on nrf52,  but I can rebase it later on after this is merged

alwa-nordic commented:
> > I removed [d8346a9](https://github.com/zephyrproject-rtos/zephyr/commit/d8346a9915f6e1d9becebc10055a7a007833f803) "Unify L2CAP and ACL bufs configuration". By the commit message, it looks like it's not needed for this PR. And it was problematic in our downstream process.
> 
> it is needed for L2CAP tests to pass on nrf52, but I can rebase it later on after this is merged

Do you know which part is needed, if we don't want to do the unification? Is it all of `tests/bluetooth/tester/prj.conf`?

alwa-nordic committed:
Bluetooth: Tester: Increase config values to make tests pass

Without this PTS tests do not pass. These values are copied from a
commit aimed at unifying the prj.conf files in this directory. I do not
know which of these are strictly necessary to make the tests pass.
Copying them all poses no harm since the unification will be done later
anyway.

Co-authored-by: Szymon Janc <szymon.janc@codecoup.pl>
Signed-off-by: Aleksander Wasaznik <aleksander.wasaznik@nordicsemi.no>

sjanc commented:
> > > I removed [d8346a9](https://github.com/zephyrproject-rtos/zephyr/commit/d8346a9915f6e1d9becebc10055a7a007833f803) "Unify L2CAP and ACL bufs configuration". By the commit message, it looks like it's not needed for this PR. And it was problematic in our downstream process.
> > 
> > 
> > it is needed for L2CAP tests to pass on nrf52, but I can rebase it later on after this is merged
> 
> Do you know which part is needed, if we don't want to do the unification? Is it all of `tests/bluetooth/tester/prj.conf`?

AFAIR it was CONFIG_BT_BUF_ACL_RX_SIZE as this affects BT_L2CAP_RX_MTU value

Thalley commented on include/zephyr/bluetooth/l2cap.h:
```diff
@@ -557,6 +558,26 @@ int bt_l2cap_ecred_chan_connect(struct bt_conn *conn,
+ *  @param chans Array of channel objects. Null-terminated. Elements after the
+ *               first 5 are silently ignored.
+ *  @param mtu Channel MTU to reconfigure to.
+ *  @param mps Channel MPS to reconfigure to.
```
```suggestion
 *  @param mps Channel MPS to reconfigure to. Value range L2CAP_ECRED_MIN_MPS to BT_L2CAP_RX_MTU
```

alwa-nordic committed:
Bluetooth: Tester: Use BT_L2CAP_SEG_RECV for L2CAP tests

This API gives better control on L2CAP COC credits and suits better
for Upper Tester implementation.

Co-authored-by: Szymon Janc <szymon.janc@codecoup.pl>
Signed-off-by: Aleksander Wasaznik <aleksander.wasaznik@nordicsemi.no>

Thalley commented on include/zephyr/bluetooth/l2cap.h:
```diff
@@ -557,6 +558,52 @@ int bt_l2cap_ecred_chan_connect(struct bt_conn *conn,
 
+/** @brief Reconfigure Enhanced Credit Based L2CAP channels
+ *
+ *  Experimental API to reconfigure L2CAP ECRED channels with explicit MPS and
+ *  MTU values.
```
Do we have a way to track experiment function APIs like this, in a way that can help us remove the experimental note later?

alwa-nordic replied:
I think so far we have promoted APIs from experimental on request. I think that makes sense. The [API lifecycle process requires multiple applications to use an API before it's promoted, to "unstable"](https://docs.zephyrproject.org/latest/develop/api/api_lifecycle.html#hardware-agnostic-apis).

Thalley replied:
Yeah, but do have any way to track individual functions? In generally the API lifecycle usually covers entire groups for functions like `bt_*`. 

We have defined versions to provide information about stable/unstable/experiemental as defined by https://docs.zephyrproject.org/latest/develop/api/overview.html, but we rarely use that for individual functions and mostly for groups of functions.

Thalley commented on include/zephyr/bluetooth/l2cap.h:
```diff
@@ -557,6 +558,52 @@ int bt_l2cap_ecred_chan_connect(struct bt_conn *conn,
+ *  Pend a L2CAP ECRED reconfiguration for up to 5 channels. All provided
+ *  channels must share the same @ref bt_conn.
+ *
+ *  Cannot decrease MTU of any channel. And, cannot decrease MPS of any channel
```
```suggestion
 *  Cannot decrease MTU of any channel, and cannot decrease MPS of any channel
```

Thalley commented on include/zephyr/bluetooth/l2cap.h:
```diff
@@ -557,6 +558,52 @@ int bt_l2cap_ecred_chan_connect(struct bt_conn *conn,
+ *  There is no dedicated callback for this operation, but whenever a peer responds to a
+ *  reconfiguration request, each channel's reconfigured() callback is invoked.
+ *
+ *  This function may block for the duration of L2CAP_RTX_TIMEOUT.
```
Yikes; this almost feel like it deserves a @warning or at least a @note. 

With the blocking changes we've seen in the system workqueue, we definitely do not want this to be called in the system workqueue then, if it can block for 2 seconds.

`L2CAP_RTX_TIMEOUT` is not a public available value. I think we should not refer to internal values like `L2CAP_RTX_TIMEOUT` in the public API. Should we move that to `l2cap.h` and `@ref` it?

alwa-nordic replied:
Maybe we should instead just say it can block for an unspecified amount of time? I think so now. Apparently `L2CAP_RTX_TIMEOUT` is implementation specific, and it looks like we chose `L2CAP_RTX_TIMEOUT` somewhat arbitrarily as the timeout for the allocation. It could change in the future.

Ideally this API should not block at all, and we should ensure this in future work. For now, this is mostly a copy-paste of `bt_l2cap_ecred_chan_reconfigure`.

Thalley replied:
> Maybe we should instead just say it can block for an unspecified amount of time? 

I think that's OK for now.

Thalley commented on include/zephyr/bluetooth/l2cap.h:
```diff
@@ -557,6 +558,52 @@ int bt_l2cap_ecred_chan_connect(struct bt_conn *conn,
+ *  @warning Known issue: The implementation returns -EBUSY if there is a
+ *  concurrent reconfigure operation on the same connection. The caller may try
+ *  again later. The event signaling readiness is on the concurrently
+ *  reconfiguring channels, which may be unavailable to the called of this
```
```suggestion
 *  reconfiguring channels, which may be unavailable to the caller of this
```

Thalley commented on include/zephyr/bluetooth/l2cap.h:
```diff
@@ -557,6 +558,52 @@ int bt_l2cap_ecred_chan_connect(struct bt_conn *conn,
+ *
+ *  This function may block for the duration of L2CAP_RTX_TIMEOUT.
+ *
+ *  @warning Known issue: The implementation returns -EBUSY if there is a
+ *  concurrent reconfigure operation on the same connection. The caller may try
```
Not sure I understand; why is this an issue? Isn't it just a fact that we can only have 1 outstanding reconfig request at any one time?

alwa-nordic replied:
The issue is that there is no event for when the reconfigure resource becomes available. I'll reword.

Thalley commented on include/zephyr/bluetooth/l2cap.h:
```diff
@@ -557,6 +558,52 @@ int bt_l2cap_ecred_chan_connect(struct bt_conn *conn,
+ *  allocate. The caller may try again later. There is no event signaling the
+ *  availability of buffers.
+ *
+ *  @warning Known issue: When more than five channels are provided, the
+ *  implementation silently ignores channels after the first five.
```
Ditto: Is this an issue, or just a fact about the API design (which I do not like :) )

alwa-nordic replied:
I would have preferred an error when too many channels are provided. I did fix that right now because I worried it could break something (by reading out-of-bounds if the caller passes an array of length 5 with no null terminator). It's was safer to keep the behavior of `bt_l2cap_ecred_chan_reconfigure`.

I should just take a look at whether Lower Tester could trigger this.

Thalley commented on include/zephyr/bluetooth/l2cap.h:
```diff
@@ -557,6 +558,52 @@ int bt_l2cap_ecred_chan_connect(struct bt_conn *conn,
+ *  @kconfig_dep{CONFIG_BT_L2CAP_RECONFIGURE_EXPLICIT}
+ *
+ *  @param chans  Null-terminated array of channels to reconfigure.
+ *  @param mtu    Desired MTU. Must be at least L2CAP_ECRED_MIN_MTU.
```
What is the maximum MTU?

alwa-nordic replied:
Good question. For channels with `seg_recv`, there is no limit, so `UINT16_MAX`. For channels with `alloc_buf`, it's the size of the buffers from `alloc_buf`, I guess? For channels with neither, it's `BT_L2CAP_SDU_RX_MTU`.

alwa-nordic replied:
We can enforce the limit when `alloc_buf` is not set. But there no good way to enforce the limit otherwise. I assume we just disconnect the channel is the application lies about the MTU. What do you think?

Thalley replied:
Sounds like that the MTU size is controlled by the upper layers then (when `alloc_buf` is set)? In that case, we cannot really limit it, and if the spec-limit is `UINT16_MAX` then we can mention that, although it is already implied by the type.

Thalley commented on include/zephyr/bluetooth/l2cap.h:
```diff
@@ -557,6 +558,52 @@ int bt_l2cap_ecred_chan_connect(struct bt_conn *conn,
+ *  @kconfig_dep{CONFIG_BT_L2CAP_RECONFIGURE_EXPLICIT}
+ *
+ *  @param chans  Null-terminated array of channels to reconfigure.
+ *  @param mtu    Desired MTU. Must be at least L2CAP_ECRED_MIN_MTU.
+ *  @param mps    Desired MPS. Must be in range L2CAP_ECRED_MIN_MPS to
```
Both `L2CAP_ECRED_MIN_MTU` and `L2CAP_ECRED_MIN_MPS` are only defined internally. Suggest to move this to the public API and `@ref` the values

sjanc commented:
#AutoPTS run zephyr L2CAP

codecoup-tester commented:
Scheduled PR https://github.com/zephyrproject-rtos/zephyr/pull/80911#issuecomment-2472747797, board: nrf52, estimated start time: 09:07:18, test case count: 93, estimated duration: 0:24:48<details><summary>Test cases to be run</summary>L2CAP/COS/CFC/BV-01-C<br>L2CAP/COS/CFC/BV-02-C<br>L2CAP/COS/CFC/BV-03-C<br>L2CAP/COS/CFC/BV-04-C<br>L2CAP/COS/CFC/BV-05-C<br>L2CAP/COS/ECFC/BV-01-C<br>L2CAP/COS/ECFC/BV-02-C<br>L2CAP/COS/ECFC/BV-03-C<br>L2CAP/COS/ECFC/BV-04-C<br>L2CAP/LE/CPU/BV-01-C<br>L2CAP/LE/CPU/BV-02-C<br>L2CAP/LE/CPU/BI-01-C<br>L2CAP/LE/CPU/BI-02-C<br>L2CAP/LE/REJ/BI-02-C<br>L2CAP/LE/CFC/BV-01-C<br>L2CAP/LE/CFC/BV-02-C<br>L2CAP/LE/CFC/BV-03-C<br>L2CAP/LE/CFC/BV-04-C<br>L2CAP/LE/CFC/BV-05-C<br>L2CAP/LE/CFC/BV-06-C<br>L2CAP/LE/CFC/BV-07-C<br>L2CAP/LE/CFC/BI-01-C<br>L2CAP/LE/CFC/BV-08-C<br>L2CAP/LE/CFC/BV-09-C<br>L2CAP/LE/CFC/BV-10-C<br>L2CAP/LE/CFC/BV-11-C<br>L2CAP/LE/CFC/BV-12-C<br>L2CAP/LE/CFC/BV-13-C<br>L2CAP/LE/CFC/BV-14-C<br>L2CAP/LE/CFC/BV-15-C<br>L2CAP/LE/CFC/BV-16-C<br>L2CAP/LE/CFC/BV-18-C<br>L2CAP/LE/CFC/BV-19-C<br>L2CAP/LE/CFC/BV-20-C<br>L2CAP/LE/CFC/BV-21-C<br>L2CAP/LE/CFC/BV-22-C<br>L2CAP/LE/CFC/BV-23-C<br>L2CAP/LE/CFC/BV-24-C<br>L2CAP/LE/CFC/BV-25-C<br>L2CAP/LE/CFC/BV-26-C<br>L2CAP/LE/CFC/BV-27-C<br>L2CAP/LE/CFC/BV-28-C<br>L2CAP/LE/CFC/BV-29-C<br>L2CAP/LE/CFC/BV-30-C<br>L2CAP/LE/CFC/BV-31-C<br>L2CAP/LE/CFC/BI-02-C<br>L2CAP/LE/CID/BI-01-C<br>L2CAP/ECFC/BV-01-C<br>L2CAP/ECFC/BV-02-C<br>L2CAP/ECFC/BV-03-C<br>L2CAP/ECFC/BV-04-C<br>L2CAP/ECFC/BV-06-C<br>L2CAP/ECFC/BV-07-C<br>L2CAP/ECFC/BI-01-C<br>L2CAP/ECFC/BI-02-C<br>L2CAP/ECFC/BV-08-C<br>L2CAP/ECFC/BV-09-C<br>L2CAP/ECFC/BV-10-C<br>L2CAP/ECFC/BV-11-C<br>L2CAP/ECFC/BV-12-C<br>L2CAP/ECFC/BV-13-C<br>L2CAP/ECFC/BV-14-C<br>L2CAP/ECFC/BV-15-C<br>L2CAP/ECFC/BV-16-C<br>L2CAP/ECFC/BV-17-C<br>L2CAP/ECFC/BV-18-C<br>L2CAP/ECFC/BV-19-C<br>L2CAP/ECFC/BV-20-C<br>L2CAP/ECFC/BV-21-C<br>L2CAP/ECFC/BV-22-C<br>L2CAP/ECFC/BV-23-C<br>L2CAP/ECFC/BI-03-C<br>L2CAP/ECFC/BV-24-C<br>L2CAP/ECFC/BV-25-C<br>L2CAP/ECFC/BI-04-C<br>L2CAP/ECFC/BV-26-C<br>L2CAP/ECFC/BV-27-C<br>L2CAP/ECFC/BV-28-C<br>L2CAP/ECFC/BV-29-C<br>L2CAP/ECFC/BV-30-C<br>L2CAP/ECFC/BV-31-C<br>L2CAP/ECFC/BV-32-C<br>L2CAP/ECFC/BV-33-C<br>L2CAP/ECFC/BV-34-C<br>L2CAP/ECFC/BV-35-C<br>L2CAP/ECFC/BV-38-C<br>L2CAP/ECFC/BV-39-C<br>L2CAP/ECFC/BV-41-C<br>L2CAP/ECFC/BI-05-C<br>L2CAP/ECFC/BI-06-C<br>L2CAP/ECFC/BI-07-C<br>L2CAP/ECFC/BI-08-C<br>L2CAP/TIM/BV-03-C</details>

codecoup-tester commented:
AutoPTS Bot results:
No failed test found.
<details><summary>Successful tests (93)</summary>L2CAP   L2CAP/COS/CFC/BV-01-C           PASS<br>L2CAP   L2CAP/COS/CFC/BV-02-C           PASS<br>L2CAP   L2CAP/COS/CFC/BV-03-C           PASS<br>L2CAP   L2CAP/COS/CFC/BV-04-C           PASS<br>L2CAP   L2CAP/COS/CFC/BV-05-C           PASS<br>L2CAP   L2CAP/COS/ECFC/BV-01-C          PASS<br>L2CAP   L2CAP/COS/ECFC/BV-02-C          PASS<br>L2CAP   L2CAP/COS/ECFC/BV-03-C          PASS<br>L2CAP   L2CAP/COS/ECFC/BV-04-C          PASS<br>L2CAP   L2CAP/ECFC/BI-01-C              PASS<br>L2CAP   L2CAP/ECFC/BI-02-C              PASS - ERRATA ES-26416<br>L2CAP   L2CAP/ECFC/BI-03-C              PASS<br>L2CAP   L2CAP/ECFC/BI-04-C              PASS<br>L2CAP   L2CAP/ECFC/BI-05-C              PASS<br>L2CAP   L2CAP/ECFC/BI-06-C              PASS<br>L2CAP   L2CAP/ECFC/BI-07-C              PASS<br>L2CAP   L2CAP/ECFC/BI-08-C              PASS<br>L2CAP   L2CAP/ECFC/BV-01-C              PASS<br>L2CAP   L2CAP/ECFC/BV-02-C              PASS<br>L2CAP   L2CAP/ECFC/BV-03-C              PASS<br>L2CAP   L2CAP/ECFC/BV-04-C              PASS<br>L2CAP   L2CAP/ECFC/BV-06-C              PASS<br>L2CAP   L2CAP/ECFC/BV-07-C              PASS<br>L2CAP   L2CAP/ECFC/BV-08-C              PASS<br>L2CAP   L2CAP/ECFC/BV-09-C              PASS<br>L2CAP   L2CAP/ECFC/BV-10-C              PASS<br>L2CAP   L2CAP/ECFC/BV-11-C              PASS<br>L2CAP   L2CAP/ECFC/BV-12-C              PASS<br>L2CAP   L2CAP/ECFC/BV-13-C              PASS<br>L2CAP   L2CAP/ECFC/BV-14-C              PASS<br>L2CAP   L2CAP/ECFC/BV-15-C              PASS<br>L2CAP   L2CAP/ECFC/BV-16-C              PASS<br>L2CAP   L2CAP/ECFC/BV-17-C              PASS<br>L2CAP   L2CAP/ECFC/BV-18-C              PASS<br>L2CAP   L2CAP/ECFC/BV-19-C              PASS<br>L2CAP   L2CAP/ECFC/BV-20-C              PASS<br>L2CAP   L2CAP/ECFC/BV-21-C              PASS<br>L2CAP   L2CAP/ECFC/BV-22-C              PASS<br>L2CAP   L2CAP/ECFC/BV-23-C              PASS<br>L2CAP   L2CAP/ECFC/BV-24-C              PASS<br>L2CAP   L2CAP/ECFC/BV-25-C              PASS<br>L2CAP   L2CAP/ECFC/BV-26-C              PASS<br>L2CAP   L2CAP/ECFC/BV-27-C              PASS<br>L2CAP   L2CAP/ECFC/BV-28-C              PASS<br>L2CAP   L2CAP/ECFC/BV-29-C              PASS<br>L2CAP   L2CAP/ECFC/BV-30-C              PASS<br>L2CAP   L2CAP/ECFC/BV-31-C              PASS<br>L2CAP   L2CAP/ECFC/BV-32-C              PASS<br>L2CAP   L2CAP/ECFC/BV-33-C              PASS<br>L2CAP   L2CAP/ECFC/BV-34-C              PASS<br>L2CAP   L2CAP/ECFC/BV-35-C              PASS<br>L2CAP   L2CAP/ECFC/BV-38-C              PASS<br>L2CAP   L2CAP/ECFC/BV-39-C              PASS<br>L2CAP   L2CAP/ECFC/BV-41-C              PASS<br>L2CAP   L2CAP/LE/CFC/BI-01-C            PASS<br>L2CAP   L2CAP/LE/CFC/BI-02-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-01-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-02-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-03-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-04-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-05-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-06-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-07-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-08-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-09-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-10-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-11-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-12-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-13-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-14-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-15-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-16-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-18-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-19-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-20-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-21-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-22-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-23-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-24-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-25-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-26-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-27-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-28-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-29-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-30-C            PASS<br>L2CAP   L2CAP/LE/CFC/BV-31-C            PASS<br>L2CAP   L2CAP/LE/CID/BI-01-C            PASS<br>L2CAP   L2CAP/LE/CPU/BI-01-C            PASS<br>L2CAP   L2CAP/LE/CPU/BI-02-C            PASS<br>L2CAP   L2CAP/LE/CPU/BV-01-C            PASS<br>L2CAP   L2CAP/LE/CPU/BV-02-C            PASS<br>L2CAP   L2CAP/LE/REJ/BI-02-C            PASS - ERRATA ES-25764<br>L2CAP   L2CAP/TIM/BV-03-C               PASS</details>

theob-pro commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -3035,6 +3039,94 @@ int bt_l2cap_ecred_chan_reconfigure(struct bt_l2cap_chan **chans, uint16_t mtu)
+		ch->pending_rx_mtu = mtu;
+
+		net_buf_add_le16(buf, ch->rx.cid);
+	};
```
![image](https://github.com/user-attachments/assets/5a93b479-f67b-48f9-adff-1fbb089066ab)

```suggestion
	}
```

Thalley commented on include/zephyr/bluetooth/l2cap.h:
```diff
@@ -83,6 +84,28 @@ extern "C" {
+ *  This requirement is inferred from text in Core 3.A.4.25 v6.0:
+ *
+ *      L2CAP implementations shall support a minimum MPS of 64 and may
+ *      support an MPS up to 65533 octets for these channels.
```
Curious; the maximum is `65533` which is not the same at UINT16_MAX. Suggest to also add a `MAX` value

alwa-nordic replied:
I was considering adding the max value define here but decided against. It would be useless in our system since we instead have a implementation defined limit `BT_L2CAP_RX_MTU` on MPS. `BT_L2CAP_RX_MTU` is limited to `65533` by kconfig.

Thalley commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -3035,6 +3039,94 @@ int bt_l2cap_ecred_chan_reconfigure(struct bt_l2cap_chan **chans, uint16_t mtu)
+
+	LOG_DBG("chans %p chan_count %u mtu 0x%04x mps 0x%04x", chans, chan_count, mtu, mps);
+
+	if (!chans || chan_count < 1 || chan_count > L2CAP_ECRED_CHAN_MAX_PER_REQ) {
```
```suggestion
	if (!chans || !IN_RANGE(chan_count, 1, L2CAP_ECRED_CHAN_MAX_PER_REQ)) {
```

Thalley replied:
We could also consider having a `L2CAP_ECRED_CHAN_MIN_PER_REQ = 1`

We could also consider moving this (these?) to the public API, as it is values that the application needs to know, and we can refer to these instead.

alwa-nordic replied:
I think 1 is obvious and does not need a define, since acting on an empty collection of channels does not make sense. We could return success instead of an error of course.

As for the maximum value, we can have a public define for that, but I think it should be renamed. This API does not say it results in exactly one request. Should it? If not then using a define with "_PER_REQ" in the define name with this API is inaccurate.

Thalley replied:
> I think 1 is obvious and does not need a define, as it just means acting on an empty collection of channels does not make sense. 

That's fine, but IMO it makes more sense to do something like 
```
value range @ref L2CAP_ECRED_CHAN_MIN_PER_REQ to @ref L2CAP_ECRED_CHAN_MAX_PER_REQ
``` 

instead of 
```
value range 1 to @ref L2CAP_ECRED_CHAN_MAX_PER_REQ
```

At least to me it is kind of odd to mix and match defined values with literal values. 

> This API does not say it results in exactly one request. Should it? If not then using a define with "_PER_REQ" in the define name with this API is inaccurate.

Not explicitly, but the requirements that it imposes (max 5 channels, cannot decrease MPS if more than one channel, etc.) indicates a single L2CAP request. 

I won't block this PR based on this comment, so OK to keep as is. Was mainly a suggestion :)

alwa-nordic replied:
There are problems with a using a define like `L2CAP_ECRED_CHAN_MIN_PER_REQ = 1` in documentation.

 1. It's an extra level of indirection. If the programmer want's to know the actual value, they have to look it up.
 2. A symbol suggests it's an abstraction, and its value could change in the future. The programmer now has to worry about that possibility.

We could avoid both pitfall and still have a define by giving it a name that reflects it's true purpose. `MIN_SIZE_OF_NONEMPTY_COLLECTION = 1`. Is the following good?

    value range @ref MIN_SIZE_OF_NONEMPTY_COLLECTION to @ref L2CAP_ECRED_CHAN_MAX_PER_REQ

Thalley replied:
You have a fair point. Looking at e.g. the ISO documentation, it does require an additional click to see the actual values of those defines: 

![image](https://github.com/user-attachments/assets/3078b7bc-6368-45f9-92ec-484d46df929e)

> A symbol suggests it's an abstraction, and its value could change in the future. The programmer now has to worry about that possibility.

That I am unsure that I agree on. It's very typical to use `#define`s for constant values that will never change. Our entire `hci_types.h` is built like that. 

> We could avoid both pitfall and still have a define by giving it a name that reflects it's true purpose. MIN_SIZE_OF_NONEMPTY_COLLECTION = 1. Is the following good?

I would prefer defined values for min and max to come from the same prefix if that makes sense. It's IMO a bit messy to do MIN from namespace A and MAX from namespace B, although I get your point. In this specific case I would actually prefer `1` over `MIN_SIZE_OF_NONEMPTY_COLLECTION`

alwa-nordic replied:
I see you already approved. To wrap up this thread: `MIN_SIZE_OF_NONEMPTY_COLLECTION` was a little absurd. I was actually expecting you to call me out on that. Instead I moved this requirement to the line above where it became the following:

```doxygen
@param chans Array of channels to reconfigure. Must be non-empty and
             contain at most 5 (@ref BT_L2CAP_ECRED_CHAN_MAX_PER_REQ)
             elements.
```

jhedberg commented on subsys/bluetooth/host/Kconfig.l2cap:
```diff
@@ -70,4 +70,12 @@ config BT_L2CAP_SEG_RECV
+	bool "L2CAP Explicit reconfigure API [EXPERIMENTAL]"
+	select EXPERIMENTAL
+	help
+
```
minor: there's normally no empty line between "help" and the actual help text.

jhedberg replied:
ah, I see this is a likely copy-paste from the `BT_L2CAP_SEG_RECV` definition. Both should be fixed.

alwa-nordic replied:
Sorry. I just have a habit of forgetting to remove the blank line after I add it to re-wrap the text with my editor. Wish we had a dedicated formatting tool for Kconfig.
